spec/index.adoc

```diff
@@ -1674,6 +1674,96 @@ Any user can top-up any canister this way.
 
 This method is only available in local development instances.
 
+
+[#ic-bitcoin-canisters]
+== The IC bitcoin canisters
+
+Similar to the <<ic-management-canister,management canister>>, the Bitcoin functionality is also exposed through a facade, which we refer to as the "Bitcoin canisters".
+There are two Bitcoin canisters differing only in that one operates on the real Bitcoin blockchain whereas the other only works with the Bitcoin testnet.
+The Bitcoin canisters can be accessed using their addresses:
+
+- The canister address of the Bitcoin canister that uses the real Bitcoin blockchain is `TBD`.
+- The canister address of the Bitcoin canister that uses the Bitcoin testnet is `TBD`.
+
+It is *not* possible to interact with either Bitcoin canister via external requests (i.e., ingress messages).
+
+Information about Bitcoin can be found in the https://developer.bitcoin.org/devguide/[Bitcoin developer guides].
+
+[#ic-bitcoin-candid]
+=== Interface overview
+
+The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the available functionality. You can also link:{attachmentsdir}/bitcoin.did[download the file].
+----
+include::{example}bitcoin.did[]
+----
+
+The binary encoding of arguments and results are as per Candid specification.
+
+[#ic-bitcoin_get_utxos]
+=== IC method `get_utxos`
+
+Given a https://en.bitcoin.it/wiki/Base58Check_encoding[base58-encoded] address as part of the
```

Comment:
Why do we use an encoded request and a `text` instead of a `blob`?

Reply:
I agree address is better typed as `blob` or even better as `principal`. `blob` is not very easy to copy/paste in the CLI. Principal has the same inner representation as blob, but is easier to copy/paste because of the textual representation and checksum. I wish we use principal for our ledger address as well.

Reply:
Alternatively, could the CLI be extended to allow the same, less error-prone, textual representation with checksum for Blobs? It seems weird to abuse Principal for something that doesn't identify an entity on the IC.

Reply:
`blob` has other use cases, e.g. the content of an asset/file. I think `Principal` isn't tied to the IC, it's a type for identity.

Reply:
Good question. I don't have a good answer personally as to why we use `text`. It was mostly chosen because that's the "standard" of how people think about addresses and provided the most ergonomic interface without any notable downsides compared to a `blob`.

I am not sure a `principal` would be better than `text`. It would have a different textual representation than bitcoin addesses, which would be very confusing for users.

Reply:
I don't think we can/should use the type principal here, since principals are senders/receivers of messages on the IC layer, and Bitcoin addresses are certainly not that. Also the support of better I/O on the CLI seems a pretty good idea (it's indeed pretty painful to use `blob` from dfx) but orthogonal to the spec.

@ielashi I think it comes down to what a Bitcoin address _is_. A principal, for instance, is an octet string of up to 29 bytes, and it has a canonical textual representation. So the question is what _is_ a Bitcoin address, and it seems you're right, is per [this](https://en.bitcoin.it/wiki/Invoice_address):
> A Bitcoin invoice address, or simply invoice, is an identifier of 26-35 alphanumeric characters, beginning with the number 1, 3 or bc1 that represents a possible destination for a bitcoin payment.
That means a Bitcoin address _is_ a `text`. I will close this thread.